modules/carbonAnalyticsAdapter.js

```diff
@@ -0,0 +1,238 @@
+      }
+
+      return localStorageId;
+    }
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(PROFILE_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      if (storage.localStorageIsEnabled()) {
+        storage.setDataInLocalStorage(PROFILE_ID_KEY, cookieId);
+      }
+
+      storage.setCookie(PROFILE_ID_COOKIE, cookieId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+
+      return cookieId;
+    }
+  }
+
+  let newId = generateUUID();
+  if (storage.localStorageIsEnabled()) {
+    storage.setDataInLocalStorage(PROFILE_ID_KEY, newId);
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    storage.setCookie(PROFILE_ID_COOKIE, newId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+  }
+
+  return newId;
+}
+
+function getSessionId() {
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(SESSION_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      storage.setCookie(SESSION_ID_COOKIE, cookieId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+      return cookieId;
+    }
+
+    let newId = generateUUID();
+    storage.setCookie(SESSION_ID_COOKIE, newId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+    return newId;
+  }
+
+  return generateUUID();
+}
+
+function registerEngagement() {
+  let present = Date.now();
+  let timediff = (present - pageEngagement.timeLastEngage) / 1000; // convert to seconds
+
+  pageEngagement.timeLastEngage = present;
+
+  if (timediff < pageEngagement.ttl) {
+    return;
+  }
+
+  pageEngagement.count++;
+  pageEngagement.id = generateUUID();
+};
+
+function getConsentData(args) {
+  let consentData = {
+    gdpr_consent: '',
+    ccpa_consent: ''
+  };
+
+  if (Array.isArray(args?.bidderRequests) && args.bidderRequests.length > 0) {
+    let bidderRequest = args.bidderRequests[0];
+
+    if (bidderRequest?.gdprConsent?.consentString) {
+      consentData.gdpr_consent = bidderRequest.gdprConsent.consentString;
+    }
+
+    if (bidderRequest?.uspConsent?.consentString) {
+      consentData.ccpa_consent = bidderRequest.uspConsent.consentString;
+    }
+  }
+
+  return consentData;
+}
+
+function getExternalIds() {
+  let externalIds = {};
+
+  if (getGlobal().getUserIdsAsEids && typeof getGlobal().getUserIdsAsEids == 'function') {
+    let eids = getGlobal().getUserIdsAsEids();
+
+    if (eids?.length) {
+      eids.forEach(eid => {
+        if (eid.source && eid?.uids?.length) {
+          externalIds[eid.source] = eid.uids.map(uid => uid.id);
+        }
+      });
+    }
+  }
+
+  return externalIds;
+}
+
+function createBaseEngagementEvent(args) {
+  let event = {};
+
+  event.profile_id = profileId;
+  event.session_id = sessionId;
+  event.pageview_id = pageViewId;
+
+  event.engagement_id = pageEngagement.id;
+  event.engagement_count = pageEngagement.count;
+  event.engagement_ttl = pageEngagement.ttl;
+
+  event.script_id = window.location.host;
+  event.url = window.location.href;
+  event.referrer = document.referrer;
```

Comment:
in my page where prebid is in top most this is always sending as empty string.

Is this wanted.

FWIW `args` when it comes from auction_end event contains a referrerInfo object at `args.bidderRequests[0].refererInfo.page` which most adapters use to set referrer.

Maybe a change here could be

```
event.referrer = document.referrer || deepAccess(args, 'bidderRequests.0.refererInfo.page') || undefined;
```

This will prefer doc referrer if set, otherwise try to grab it from prebids referer resultion, else set to undefined which will not send it.